### ISSUE TITLE
normal devices get marked as hidden on older sqlite versions

### ISSUE THREAD

**Issue by uhoreg** [matrix-org/synapse#6302]
e.g. sqlite3 from Debian stretch.  Because the table schema sets the default value to `FALSE`, which causes older sqlite3 versions to default to the string 'FALSE' which is truthy.

**Comment by micyp** [comment]
To allow affected users to find this issue, I'm posting the error message that is produced in the log:

`2019-10-30 17:39:50,484 - synapse.handlers.e2e_keys - 434 - INFO - POST-114- Updating device_keys for device 'REDACTED' for user @REDACTED:REDACTED at 1572453590483`
`2019-10-30 17:39:50,488 - synapse.http.server - 80 - INFO - POST-114- <XForwardedForRequest at 0x7f3200228278 method='POST' uri='/_matrix/client/unstable/keys/upload/REDACTED' clientproto='HTTP/1.0' site=8008> SynapseError: 400 - The device ID is in use`
`2019-10-30 17:39:50,489 - synapse.access.http.8008 - 302 - INFO - POST-114- REDACTED - 8008 - {@REDACTED:REDACTED} Processed request: 0.006sec/0.000sec (0.004sec, 0.000sec) (0.001sec/0.001sec/3) 59B 400 "POST /_matrix/client/unstable/keys/upload/REDACTED HTTP/1.0" "Riot.im/0.9.8 (Linux; U; Android 7.1.2; REDACTED; Flavour FDroid; MatrixAndroidSDK 0.9.30)" [0 dbevts]
`

To work around this issue, run the following SQL statement on your database:
`UPDATE devices SET hidden = 0 WHERE hidden = 'FALSE'`
(credit @uhoreg)

**Comment by uhoreg** [comment]
It seems like [sqlite 3.23.0](https://sqlite.org/releaselog/3_23_0.html) is when the behaviour changed and it recognized `TRUE` and `FALSE` properly.

I think the best solution is to change the schema for sqlite to be `... hidden BOOLEAN DEFAULT 0`, and keep as-is for postgresql.  (We can't change the schema for postgresql because it won't accept a default of 0 for a boolean field.)

**Comment by schildbach** [comment]
I'm bitten by this bug, too. Here is the respective logfile:

```
2019-11-07 08:53:04,189 - synapse.access.http.8008 - 233 - INFO - POST-54- [REDACTED] - 8008 - Received request: POST /_matrix/client/api/v1/login
2019-11-07 08:53:04,191 - synapse.rest.client.v1.login - 176 - INFO - POST-54- Got login request with identifier: None, medium: None, address: None, user: '[REDACTED]'
2019-11-07 08:53:04,655 - synapse.http.server - 80 - INFO - POST-54- <XForwardedForRequest at 0x7fb8554e7748 method='POST' uri='/_matrix/client/api/v1/login' clientproto='HTTP/1.0' site=8008> SynapseError: 400 - The device ID is in use
2019-11-07 08:53:04,657 - synapse.access.http.8008 - 302 - INFO - POST-54- [REDACTED] - 8008 - {None} Processed request: 0.467sec/0.001sec (0.411sec, 0.036sec) (0.003sec/0.001sec/3) 59B 400 "POST /_matrix/client/api/v1/login HTTP/1.0" "-" [0 dbevts]
```

**Comment by richvdh** [comment]
fixed by #6313

**Comment by schildbach** [comment]
I can assert that `UPDATE devices SET hidden = 0 WHERE hidden = 'FALSE'` made my node operable again. However I think new devices would still be created with the wrong `hidden` value since the default hasn't changed. When will the patch be rolled out?
